File: src/lib/revenue.ts
import { db } from "./db";
import { shiftMonth, todayISO } from "./utils";

/** Libya is UTC+2 year-round. */
export function monthBounds(month: string) {
  const safe = /^\d{4}-\d{2}$/.test(month) ? month : todayISO().slice(0, 7);
  const next = shiftMonth(safe, 1);
  return {
    month: safe,
    start: new Date(`${safe}-01T00:00:00+02:00`),
    end: new Date(`${next}-01T00:00:00+02:00`),
    next,
  };
}

export async function availableMonths() {
  const current = todayISO().slice(0, 7);
  const [firstFee, firstPay] = await Promise.all([
    db.platformFee.findFirst({ orderBy: { createdAt: "asc" }, select: { createdAt: true } }),
    db.subscriptionPayment.findFirst({
      where: { status: "CONFIRMED", reviewedAt: { not: null } },
      orderBy: { reviewedAt: "asc" },
      select: { reviewedAt: true },
    }),
  ]);
  const starts = [firstFee?.createdAt, firstPay?.reviewedAt].filter(Boolean) as Date[];
  const earliest = starts.length
    ? starts.reduce((min, d) => (d < min ? d : min)).toISOString().slice(0, 7)
    : current;
  const months: string[] = [];
  let cursor = earliest;
  while (cursor <= current) {
    months.push(cursor);
    cursor = shiftMonth(cursor, 1);
  }
  return months.reverse();
}

export async function revenueForMonth(month: string) {
  const { start, end } = monthBounds(month);
  const [feeGroups, payGroups, outstandingGroups] = await Promise.all([
    db.platformFee.groupBy({
      by: ["artistId"],
      where: { createdAt: { gte: start, lt: end } },
      _sum: { amountLyd: true },
      _count: { _all: true },
    }),
    db.subscriptionPayment.groupBy({
      by: ["artistId"],
      where: { status: "CONFIRMED", reviewedAt: { gte: start, lt: end } },
      _sum: { amountLyd: true },
      _count: { _all: true },
    }),
    db.platformFee.groupBy({
      by: ["artistId"],
      where: { status: "UNPAID" },
      _sum: { amountLyd: true },
    }),
  ]);

  const ids = [...new Set([...feeGroups.map((g) => g.artistId), ...payGroups.map((g) => g.artistId)])];
  const artists = ids.length
    ? await db.artist.findMany({
        where: { id: { in: ids } },
        select: { id: true, name: true, slug: true, neighborhood: true },
      })
    : [];
  const byId = Object.fromEntries(artists.map((a) => [a.id, a]));
  const payById = Object.fromEntries(payGroups.map((g) => [g.artistId, g]));
  const dueById = Object.fromEntries(outstandingGroups.map((g) => [g.artistId, g._sum.amountLyd || 0]));

  const rows = feeGroups
    .map((g) => {
      const artist = byId[g.artistId];
      const collected = payById[g.artistId]?._sum.amountLyd || 0;
      return {
        artistId: g.artistId,
        name: artist?.name || "خبيرة",
        slug: artist?.slug || "",
        neighborhood: artist?.neighborhood || "",
        generatedLyd: g._sum.amountLyd || 0,
        bookingCount: g._count._all,
        collectedLyd: collected,
        outstandingLyd: dueById[g.artistId] || 0,
      };
    })
    .concat(
      payGroups
        .filter((g) => !feeGroups.some((f) => f.artistId === g.artistId))
        .map((g) => {
          const artist = byId[g.artistId];
          return {
            artistId: g.artistId,
            name: artist?.name || "خبيرة",
            slug: artist?.slug || "",
            neighborhood: artist?.neighborhood || "",
            generatedLyd: 0,
            bookingCount: 0,
            collectedLyd: g._sum.amountLyd || 0,
            outstandingLyd: dueById[g.artistId] || 0,
          };
        }),
    )
    .sort((a, b) => b.generatedLyd - a.generatedLyd || b.collectedLyd - a.collectedLyd);

  const generatedLyd = rows.reduce((sum, row) => sum + row.generatedLyd, 0);
  const collectedLyd = rows.reduce((sum, row) => sum + row.collectedLyd, 0);
  const bookingCount = rows.reduce((sum, row) => sum + row.bookingCount, 0);
  const top = rows[0] || null;

  return {
    month,
    generatedLyd,
    collectedLyd,
    bookingCount,
    artistCount: rows.length,
    collectionRate: generatedLyd ? Math.round((collectedLyd / generatedLyd) * 100) : collectedLyd ? 100 : 0,
    topArtist: top
      ? {
          artistId: top.artistId,
          name: top.name,
          slug: top.slug,
          generatedLyd: top.generatedLyd,
          share: generatedLyd ? Math.round((top.generatedLyd / generatedLyd) * 100) : 0,
        }
      : null,
    artists: rows.map((row, index) => ({
      ...row,
      rank: index + 1,
      share: generatedLyd ? Math.round((row.generatedLyd / generatedLyd) * 100) : 0,
    })),
  };
}

export async function revenueTrend(endMonth: string, count = 6) {
  const months: string[] = [];
  for (let i = count - 1; i >= 0; i -= 1) months.push(shiftMonth(endMonth, -i));
  const points = [];
  for (const month of months) {
    const { start, end } = monthBounds(month);
    const [generated, collected] = await Promise.all([
      db.platformFee.aggregate({
        where: { createdAt: { gte: start, lt: end } },
        _sum: { amountLyd: true },
        _count: true,
      }),
      db.subscriptionPayment.aggregate({
        where: { status: "CONFIRMED", reviewedAt: { gte: start, lt: end } },
        _sum: { amountLyd: true },
      }),
    ]);
    points.push({
      month,
      generatedLyd: generated._sum.amountLyd || 0,
      collectedLyd: collected._sum.amountLyd || 0,
      bookingCount: generated._count,
    });
  }
  return points;
}

File: backend/bridey_api/revenue.py
from __future__ import annotations

import re
from datetime import datetime, timezone

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from bridey_api.dates import shift_month, today_iso
from bridey_api.models import Artist, PlatformFee, SubscriptionPayment


def month_bounds(month: str) -> dict:
    safe = month if re.fullmatch(r"\d{4}-\d{2}", month or "") else today_iso()[:7]
    nxt = shift_month(safe, 1)
    start = datetime.fromisoformat(f"{safe}-01T00:00:00+02:00").astimezone(timezone.utc).replace(tzinfo=None)
    end = datetime.fromisoformat(f"{nxt}-01T00:00:00+02:00").astimezone(timezone.utc).replace(tzinfo=None)
    return {"month": safe, "start": start, "end": end, "next": nxt}


def available_months(session: Session) -> list[str]:
    current = today_iso()[:7]
    first_fee = session.scalar(select(PlatformFee.created_at).order_by(PlatformFee.created_at.asc()))
    first_pay = session.scalar(
        select(SubscriptionPayment.reviewed_at)
        .where(SubscriptionPayment.status == "CONFIRMED", SubscriptionPayment.reviewed_at.is_not(None))
        .order_by(SubscriptionPayment.reviewed_at.asc())
    )
    starts = [value for value in (first_fee, first_pay) if value]
    if starts:
        earliest_dt = min(starts)
        earliest = earliest_dt.isoformat()[:7]
    else:
        earliest = current
    months: list[str] = []
    cursor = earliest
    while cursor <= current:
        months.append(cursor)
        cursor = shift_month(cursor, 1)
    months.reverse()
    return months


def revenue_for_month(session: Session, month: str) -> dict:
    bounds = month_bounds(month)
    start, end = bounds["start"], bounds["end"]
    fee_rows = session.execute(
        select(PlatformFee.artist_id, func.sum(PlatformFee.amount_lyd), func.count())
        .where(PlatformFee.created_at >= start, PlatformFee.created_at < end)
        .group_by(PlatformFee.artist_id)
    ).all()
    pay_rows = session.execute(
        select(SubscriptionPayment.artist_id, func.sum(SubscriptionPayment.amount_lyd), func.count())
        .where(
            SubscriptionPayment.status == "CONFIRMED",
            SubscriptionPayment.reviewed_at >= start,
            SubscriptionPayment.reviewed_at < end,
        )
        .group_by(SubscriptionPayment.artist_id)
    ).all()
    due_rows = session.execute(
        select(PlatformFee.artist_id, func.sum(PlatformFee.amount_lyd))
        .where(PlatformFee.status == "UNPAID")
        .group_by(PlatformFee.artist_id)
    ).all()
    ids = list({row[0] for row in fee_rows} | {row[0] for row in pay_rows})
    artists = session.scalars(select(Artist).where(Artist.id.in_(ids))).all() if ids else []
    by_id = {artist.id: artist for artist in artists}
    pay_by_id = {row[0]: row for row in pay_rows}
    due_by_id = {row[0]: int(row[1] or 0) for row in due_rows}
    rows = []
    fee_ids = {row[0] for row in fee_rows}
    for artist_id, amount, count in fee_rows:
        artist = by_id.get(artist_id)
        collected = int(pay_by_id.get(artist_id, (None, 0, 0))[1] or 0)
        rows.append(
            {
                "artistId": artist_id,
                "name": artist.name if artist else "خبيرة",
                "slug": artist.slug if artist else "",
                "neighborhood": artist.neighborhood if artist else "",
                "generatedLyd": int(amount or 0),
                "bookingCount": int(count or 0),
                "collectedLyd": collected,
                "outstandingLyd": due_by_id.get(artist_id) or 0,
            }
        )
    for artist_id, amount, _count in pay_rows:
        if artist_id in fee_ids:
            continue
        artist = by_id.get(artist_id)
        rows.append(
            {
                "artistId": artist_id,
                "name": artist.name if artist else "خبيرة",
                "slug": artist.slug if artist else "",
                "neighborhood": artist.neighborhood if artist else "",
                "generatedLyd": 0,
                "bookingCount": 0,
                "collectedLyd": int(amount or 0),
                "outstandingLyd": due_by_id.get(artist_id) or 0,
            }
        )
    rows.sort(key=lambda row: (-row["generatedLyd"], -row["collectedLyd"]))
    generated = sum(row["generatedLyd"] for row in rows)
    collected = sum(row["collectedLyd"] for row in rows)
    booking_count = sum(row["bookingCount"] for row in rows)
    top = rows[0] if rows else None
    return {
        "month": month,
        "generatedLyd": generated,
        "collectedLyd": collected,
        "bookingCount": booking_count,
        "artistCount": len(rows),
        "collectionRate": round((collected / generated) * 100) if generated else (100 if collected else 0),
        "topArtist": {
            "artistId": top["artistId"],
            "name": top["name"],
            "slug": top["slug"],
            "generatedLyd": top["generatedLyd"],
            "share": round((top["generatedLyd"] / generated) * 100) if generated else 0,
        }
        if top
        else None,
        "artists": [
            {
                **row,
                "rank": index + 1,
                "share": round((row["generatedLyd"] / generated) * 100) if generated else 0,
            }
            for index, row in enumerate(rows)
        ],
    }


def revenue_trend(session: Session, end_month: str, count: int = 6) -> list[dict]:
    months = [shift_month(end_month, -i) for i in range(count - 1, -1, -1)]
    points = []
    for month in months:
        bounds = month_bounds(month)
        generated = session.execute(
            select(func.coalesce(func.sum(PlatformFee.amount_lyd), 0), func.count())
            .where(PlatformFee.created_at >= bounds["start"], PlatformFee.created_at < bounds["end"])
        ).one()
        collected = session.scalar(
            select(func.coalesce(func.sum(SubscriptionPayment.amount_lyd), 0)).where(
                SubscriptionPayment.status == "CONFIRMED",
                SubscriptionPayment.reviewed_at >= bounds["start"],
                SubscriptionPayment.reviewed_at < bounds["end"],
            )
        )
        points.append(
            {
                "month": month,
                "generatedLyd": int(generated[0] or 0),
                "collectedLyd": int(collected or 0),
                "bookingCount": int(generated[1] or 0),
            }
        )
    return points
